File: quantum_cam_rng.py
"""quantum_cam_rng.py
=====================================
A webcam‑based true‑random‑number generator that mirrors much of the
``random`` standard‑library API while always returning **NumPy arrays**.

Highlights (v0.9)
-----------------
* **NEW**: Discards pixel values outside the usable intensity window
  ``15 ≤ val ≤ 240`` before extracting their least‑significant bits.
* Flattened RGB‑interleaved bitstream (R,G,B, R,G,B, …).
* ``patch_size`` counts **pixels** (e.g., 100 → 300 bits/patch).
* XOR‑reduced (default) or ``full_patch_bits=True`` to keep every bit.
* Helpers: ``randint``, ``randrange``, ``random``, ``choice``,
  ``choices``, ``shuffle``, ``sample``.
* Demo at the bottom exercises every helper.
"""
from __future__ import annotations

from functools import lru_cache
from typing import Sequence, Tuple, Union, List

import cv2
import numpy as _np

# ---------------------------------------------------------------------------
# Optional GPU backend
# ---------------------------------------------------------------------------
try:
    import cupy as _cupy  # type: ignore

    _xp = _cupy  # type: ignore
    _USING_GPU = True
except ImportError:  # CPU fallback
    _xp = _np  # type: ignore
    _USING_GPU = False

# ---------------------------------------------------------------------------
# Backend → NumPy helper
# ---------------------------------------------------------------------------

def _as_numpy(arr):
    """Return *arr* as NumPy, copying from GPU if necessary."""
    if _USING_GPU and isinstance(arr, _xp.ndarray):  # type: ignore
        return _xp.asnumpy(arr)  # type: ignore[attr-defined]
    return _np.asarray(arr)

# ---------------------------------------------------------------------------
# GPU XOR helper (CuPy lacks reduce for xor)
# ---------------------------------------------------------------------------

def _xor_reduce_gpu(arr: "_cupy.ndarray") -> "_cupy.ndarray":  # noqa: F821
    res = arr[:, 0].copy()
    for i in range(1, arr.shape[1]):
        res ^= arr[:, i]
    return res

@lru_cache(maxsize=1)
def _powers_of_two(dtype=_xp.uint64):
    return _xp.power(_xp.array(2, dtype=dtype), _xp.arange(63, -1, -1, dtype=dtype))

# ---------------------------------------------------------------------------
# RNG class
# ---------------------------------------------------------------------------

class QuantumCamRNG:
    """Static‑style RNG backed by a webcam sensor."""

    # Configurable class‑wide settings
    _patch_size: int = 50         # pixels per patch
    _webcam_index: int = 0
    _full_patch_bits: bool = False

    # Internal camera singleton
    _camera: "QuantumCamRNG._Camera | None" = None

    # Expose backend details
    xp = _xp
    _using_gpu = _USING_GPU

    # ------------------------------------------------------------------
    # Constructor to update class defaults
    # ------------------------------------------------------------------
    @classmethod
    def init(cls, *, patch_size: int | None = None, webcam_index: int | None = None, full_patch_bits: bool | None = None):
        if patch_size is not None:
            cls._patch_size = int(patch_size)
        if full_patch_bits is not None:
            cls._full_patch_bits = bool(full_patch_bits)
        if webcam_index is not None and webcam_index != cls._webcam_index:
            # Switch camera → close old one first
            cls.close()
            cls._webcam_index = int(webcam_index)

    # ------------------------------------------------------------------
    # Camera wrapper
    # ------------------------------------------------------------------
    class _Camera:
        """Wrapper that keeps a re‑usable interleave buffer."""

        # Intensity window constants
        _MIN_VAL = 15
        _MAX_VAL = 240

        def __init__(self, index: int):
            self.cap = cv2.VideoCapture(index)
            if not self.cap.isOpened():
                raise RuntimeError(f"Cannot open webcam (index {index})")
            self._buf = None  # lazily allocated interleave buffer

        def close(self):
            if self.cap.isOpened():
                self.cap.release()

        def frame_bits(self, cls: "type[QuantumCamRNG]") -> "_xp.ndarray":  # noqa: F821
            """Read one frame and return an interleaved 1‑bit RGB stream.

            Pixels with any channel outside the [15, 240] range are *ignored*.
            The three colour channels are kept in lock‑step; if the usable
            counts differ the shortest channel length determines the batch.
            """
            ok, frame_bgr = self.cap.read()
            if not ok:
                raise RuntimeError("Failed to read frame from webcam")

            frame = cls.xp.asarray(frame_bgr, dtype=cls.xp.uint8)

            # Split channels & create intensity masks (per channel)
            r_chan = frame[:, :, 2]
            g_chan = frame[:, :, 1]
            b_chan = frame[:, :, 0]

            r_mask = (r_chan >= self._MIN_VAL) & (r_chan <= self._MAX_VAL)
            g_mask = (g_chan >= self._MIN_VAL) & (g_chan <= self._MAX_VAL)
            b_mask = (b_chan >= self._MIN_VAL) & (b_chan <= self._MAX_VAL)

            # Extract LSBs as uint8 vectors
            r_bits = (r_chan & 1).ravel()
            g_bits = (g_chan & 1).ravel()
            b_bits = (b_chan & 1).ravel()

            # Apply masks & keep equal count across channels
            r_filtered = r_bits[r_mask.ravel()]
            g_filtered = g_bits[g_mask.ravel()]
            b_filtered = b_bits[b_mask.ravel()]

            n_valid = int(cls.xp.minimum(cls.xp.minimum(r_filtered.size, g_filtered.size), b_filtered.size))
            if n_valid == 0:
                # No sufficiently exposed pixels this frame
                return cls.xp.empty(0, dtype=cls.xp.uint8)

            r_filtered = r_filtered[:n_valid]
            g_filtered = g_filtered[:n_valid]
            b_filtered = b_filtered[:n_valid]

            # Prepare / reuse interleave buffer
            required = n_valid * 3
            if self._buf is None or self._buf.size != required:
                self._buf = cls.xp.empty(required, dtype=cls.xp.uint8)
            inter = self._buf
            inter[0::3], inter[1::3], inter[2::3] = r_filtered, g_filtered, b_filtered
            return inter

    # ------------------------------------------------------------------
    # Helper getters / processors
    # ------------------------------------------------------------------
    @classmethod
    def _cam(cls):
        if cls._camera is None:
            cls._camera = cls._Camera(cls._webcam_index)
        return cls._camera

    @classmethod
    def _patch_process(cls, flat_bits: "_xp.ndarray") -> "_xp.ndarray":  # noqa: F821
        bits_per_patch = cls._patch_size * 3
        if cls._full_patch_bits:
            return flat_bits
        n = flat_bits.size // bits_per_patch
        if n == 0:
            return cls.xp.empty(0, dtype=cls.xp.uint8)
        trimmed = flat_bits[: n * bits_per_patch].reshape(n, bits_per_patch)
        if cls._using_gpu:
            return _xor_reduce_gpu(trimmed)  # type: ignore[arg-type]
        return _np.bitwise_xor.reduce(trimmed, axis=1)  # type: ignore[arg-type]

    @classmethod
    def _bits_to_uint64(cls, bits: "_xp.ndarray") -> "_xp.ndarray":  # noqa: F821
        n = bits.size // 64
        if n == 0:
            return cls.xp.empty(0, dtype=cls.xp.uint64)
        shaped = bits[: n * 64].reshape(n, 64)
        return cls.xp.dot(shaped.astype(cls.xp.uint64), _powers_of_two())

    @staticmethod
    def _u64_to_float32(u64: "_xp.ndarray") -> _np.ndarray:  # noqa: F821
        max_u64 = _np.iinfo(_np.uint64).max
        return (_as_numpy(u64).astype(_np.float32) / max_u64).astype(_np.float32)

    # ------------------------------------------------------------------
    # Bit / float collectors (always NumPy)
    # ------------------------------------------------------------------
    @classmethod
    def _collect_bits(cls, n: int) -> _np.ndarray:
        out: List[_np.ndarray] = []
        total = 0
        while total < n:
            bits = cls._patch_process(cls._cam().frame_bits(cls))
            if bits.size == 0:
                continue  # skip empty frames
            arr = _as_numpy(bits).astype(_np.uint8, copy=False)
            out.append(arr)
            total += arr.size
        return _np.concatenate(out)[:n]

    @classmethod
    def _collect_floats(cls, n: int) -> _np.ndarray:
        out: List[_np.ndarray] = []
        total = 0
        while total < n:
            bits = cls._patch_process(cls._cam().frame_bits(cls))
            if bits.size == 0:
                continue
            out.append(cls._u64_to_float32(cls._bits_to_uint64(bits)))
            total += out[-1].size
        return _np.concatenate(out)[:n].astype(_np.float32, copy=False)

    # ------------------------------------------------------------------
    # Public numeric API
    # ------------------------------------------------------------------
    @classmethod
    def random(cls, size: Union[int, Sequence[int], Tuple[int, ...]] = 1) -> _np.ndarray:
        if isinstance(size, int):
            need, shape = size, (size,)
        else:
            shape = tuple(size)  # type: ignore[arg-type]
            need = int(_np.prod(shape))
        return cls._collect_floats(need).reshape(shape)

    @classmethod
    def randint(cls, low: int, high: int | None = None, size: Union[int, Sequence[int], Tuple[int, ...]] = 1, dtype=_np.int32,) -> _np.ndarray | int:
        if high is None:
            low, high = 0, low
        if high <= low:
            raise ValueError("high must be > low")
        if isinstance(size, int):
            need, shape = size, (size,)
        else:
            shape = tuple(size)  # type: ignore[arg-type]
            need = int(_np.prod(shape))
        floats = cls._collect_floats(need)
        ints = ((floats * (high - low)).astype(dtype) + low).reshape(shape)
        return ints if need != 1 else ints.item()

    @classmethod
    def randrange(cls, start: int, stop: int | None = None, step: int = 1) -> int:
        if step == 0:
            raise ValueError("step must not be zero")
        if stop is None:
            start, stop = 0, start
        n = (stop - start + (step - 1)) // step
        if n <= 0:
            raise ValueError("empty range")
        k = int(cls.randint(0, n))
        return start + k * step
    
    @classmethod
    def choice(cls, seq: Sequence):
        if not seq:
            raise IndexError("Cannot choose from an empty sequence")
        idx = int(cls.randint(0, len(seq)))
        return seq[idx]

    @classmethod
    def choices(cls, seq: Sequence, k: int = 1) -> List:
        if not seq:
            raise IndexError("Cannot choose from an empty sequence")
        idxs = cls.randint(0, len(seq), size=k).tolist()
        return [seq[i] for i in idxs]

    @classmethod
    def shuffle(cls, seq: list) -> None:
        n = len(seq)
        for i in range(n - 1, 0, -1):
            j = int(cls.randint(0, i + 1))
            seq[i], seq[j] = seq[j], seq[i]

    @classmethod
    def sample(cls, seq: Sequence, k: int) -> List:
        if k < 0 or k > len(seq):
            raise ValueError("Sample larger than population or negative")
        idxs = list(range(len(seq)))
        cls.shuffle(idxs)  # in‑place Fisher–Yates on the indices
        return [seq[i] for i in idxs[:k]]

    # ------------------------------------------------------------------
    # Clean‑up
    # ------------------------------------------------------------------
    @classmethod
    def close(cls):
        if cls._camera is not None:
            cls._camera.close()
            cls._camera = None


RNG = QuantumCamRNG

__all__ = ["QuantumCamRNG", "RNG", "main"]

# ---------------------------------------------------------------------------
# Demo
# ---------------------------------------------------------------------------

def main():
    """Run a rich demonstration of the RNG helpers (Ctrl+C to quit)."""
    print("Quantum‑camera RNG demo – Ctrl+C to quit")
    try:
        # Configure: 100‑pixel patches (300 bits) in XOR mode initially
        RNG.init(patch_size=10)

        # --- Frame statistics ------------------------------------------------
        cam = RNG._cam()
        raw_bits = cam.frame_bits(RNG)
        xor_bits = RNG._patch_process(raw_bits)
        print("Raw interleaved bits in frame:", len(raw_bits))
        print("XOR‑mode bits in frame:", xor_bits.size)

        # Switch to full‑patch harvesting and show new bit count
        RNG.init(full_patch_bits=True)
        full_bits = RNG._patch_process(raw_bits)
        print("Full‑patch bits in frame:", full_bits.size)

        # --- Numeric helper showcase ----------------------------------------
        print("Helper function examples:")
        print("randint(1, 100):", int(RNG.randint(1, 101)))  # inclusive high bound
        print("randrange(1, 100, 5):", RNG.randrange(1, 100, 5))

        seq = [10, 20, 30, 40, 50]
        print("choice(seq):", RNG.choice(seq))
        print("choices(seq, k=3):", RNG.choices(seq, k=3))
        print("sample(seq, k=2):", RNG.sample(seq, k=2))

        to_shuffle = [1, 2, 3, 4, 5]
        RNG.shuffle(to_shuffle)
        print("shuffle([1,2,3,4,5]):", to_shuffle)

    finally:
        RNG.close()
        print("Camera closed.")


if __name__ == "__main__":
    main()

File: quantum_cam_rng.js
/* quantum_cam_rng.js – v0.2 (browser‑only)
   ------------------------------------------------------
   Fixes:
   • Removed signed‑shift overflow in `_bitsToFloat32` (no more negative
     floats ⇒ `randint` cannot yield negatives; shuffle/sample no longer
     produce `null`).
   • Minor doc and safety tweaks.
*/

export class QuantumCamRNG {
  /* ------------------ construction & camera ------------------ */
  constructor({ patchSize = 50, fullPatchBits = false, videoConstraints = { video: { width: 640, height: 480 } } } = {}) {
    this.patchSize = patchSize;           // pixels per patch
    this.fullPatchBits = fullPatchBits;   // keep every bit vs XOR‑reduce
    this.constraints = videoConstraints;  // passed to getUserMedia

    this._stream = null;          // MediaStream from getUserMedia
    this._video = null;           // <video> element (hidden)
    this._canvas = null;          // off‑screen <canvas>
    this._ctx = null;             // 2D context
    this._buf = null;             // Uint8Array interleave buffer

    this._MIN_VAL = 15;
    this._MAX_VAL = 240;
  }

  /* Start webcam & allocate helpers.  Returns **this** once ready. */
  async start() {
    if (this._stream) return this; // already started

    // 1) Request camera
    this._stream = await navigator.mediaDevices.getUserMedia(this.constraints);

    // 2) Hidden video sink
    this._video = document.createElement("video");
    this._video.style.display = "none";
    this._video.srcObject = this._stream;
    this._video.playsInline = true;
    await this._video.play(); // wait until metadata loaded & playback begun

    // 3) Off‑screen canvas matching video size
    this._canvas = document.createElement("canvas");
    this._canvas.width = this._video.videoWidth;
    this._canvas.height = this._video.videoHeight;
    this._ctx = this._canvas.getContext("2d", { willReadFrequently: true });

    return this;
  }

  /* Stop camera & release resources. */
  stop() {
    if (this._stream) {
      this._stream.getTracks().forEach(t => t.stop());
      this._stream = null;
    }
    this._video?.remove();
    this._video = null;
    this._canvas = null;
    this._ctx = null;
    this._buf = null;
  }

  /* ------------------ low‑level helpers ------------------ */
  /* Capture a single frame; return **Uint8Array** of interleaved bits. */
  _frameBits() {
    // Draw current video frame on canvas
    this._ctx.drawImage(this._video, 0, 0, this._canvas.width, this._canvas.height);
    const { data } = this._ctx.getImageData(0, 0, this._canvas.width, this._canvas.height);
    // data is Uint8ClampedArray in RGBA order

    // Allocate / reuse buffer for interleaved bits (RGB only)
    const nPix = data.length >>> 2; // divide by 4 using bitwise
    if (!this._buf || this._buf.length !== nPix * 3) {
      this._buf = new Uint8Array(nPix * 3);
    }

    let count = 0; // number of **usable** pixels
    for (let i = 0; i < data.length; i += 4) {
      const r = data[i];
      const g = data[i + 1];
      const b = data[i + 2];
      if (r < this._MIN_VAL || r > this._MAX_VAL) continue;
      if (g < this._MIN_VAL || g > this._MAX_VAL) continue;
      if (b < this._MIN_VAL || b > this._MAX_VAL) continue;

      this._buf[count * 3] = r & 1;
      this._buf[count * 3 + 1] = g & 1;
      this._buf[count * 3 + 2] = b & 1;
      count += 1;
    }
    return this._buf.subarray(0, count * 3); // trim to valid content
  }

  /* XOR‑reduce patches unless fullPatchBits=true. */
  _patchProcess(bits) {
    if (this.fullPatchBits) return bits;
    const bitsPerPatch = this.patchSize * 3;
    const n = Math.floor(bits.length / bitsPerPatch);
    if (n === 0) return new Uint8Array(0);

    const out = new Uint8Array(n);
    for (let i = 0; i < n; i++) {
      let acc = 0;
      const offset = i * bitsPerPatch;
      for (let j = 0; j < bitsPerPatch; j++) acc ^= bits[offset + j];
      out[i] = acc;
    }
    return out;
  }

  /* Collect ≥ n bits (Uint8Array) – may span multiple frames. */
  async _collectBits(n) {
    const chunks = [];
    let total = 0;
    while (total < n) {
      await new Promise(requestAnimationFrame); // yield to next frame
      const bits = this._patchProcess(this._frameBits());
      if (bits.length === 0) continue; // skip empty frame
      chunks.push(bits);
      total += bits.length;
    }
    const out = new Uint8Array(n);
    let pos = 0;
    for (const c of chunks) {
      out.set(c.subarray(0, Math.min(c.length, n - pos)), pos);
      pos += c.length;
      if (pos >= n) break;
    }
    return out;
  }

  /* Convert 32 LSB bits (Uint8Array) → float ∈ [0,1).  (unsigned!) */
  _bitsToFloat32(bits32) {
    let val = 0;
    for (let i = 0; i < 32; i++) {
      // Use arithmetic not bitwise to avoid signed 32‑bit overflow.
      val = val * 2 + bits32[i];
    }
    return val / 4294967296; // 2^32
  }

  /* ------------- public API ------------- */
  /* random(k)  → Float32Array of k uniform floats in [0,1) */
  async random(k = 1) {
    const needBits = k * 32;
    const bits = await this._collectBits(needBits);
    const out = new Float32Array(k);
    for (let i = 0; i < k; i++) {
      out[i] = this._bitsToFloat32(bits.subarray(i * 32, (i + 1) * 32));
    }
    return k === 1 ? out[0] : out;
  }

  /* randint(low, high, k=1) – high **exclusive** (Python‑style). */
  async randint(low, high = null, k = 1) {
    if (high === null) { high = low; low = 0; }
    if (high <= low) throw new RangeError("high must be > low");
    const range = high - low;
    const floats = await this.random(k);
    const arrayified = k === 1 ? [floats] : Array.from(floats);
    const ints = arrayified.map(f => low + Math.floor(f * range));
    return k === 1 ? ints[0] : ints;
  }

  /* randrange(start, stop, step=1) – single int. */
  async randrange(start, stop = null, step = 1) {
    if (step === 0) throw new RangeError("step must not be zero");
    if (stop === null) { stop = start; start = 0; }
    const n = Math.floor((stop - start + step - 1) / step);
    if (n <= 0) throw new RangeError("empty range");
    const idx = await this.randint(0, n);
    return start + idx * step;
  }

  /* choice(array) */
  async choice(arr) {
    if (!arr.length) throw new RangeError("Cannot choose from an empty array");
    const idx = await this.randint(0, arr.length);
    return arr[idx];
  }

  /* choices(array, k) – with replacement */
  async choices(arr, k = 1) {
    if (!arr.length) throw new RangeError("Cannot choose from an empty array");
    const idxs = await this.randint(0, arr.length, k);
    return idxs.map(i => arr[i]);
  }

  /* shuffle(array) – in‑place Fisher‑Yates */
  async shuffle(arr) {
    for (let i = arr.length - 1; i > 0; i--) {
      const j = await this.randint(0, i + 1);
      [arr[i], arr[j]] = [arr[j], arr[i]];
    }
  }

  /* sample(array, k) – without replacement */
  async sample(arr, k) {
    if (k < 0 || k > arr.length) throw new RangeError("Sample larger than population or negative");
    const pool = arr.slice();
    await this.shuffle(pool);
    return pool.slice(0, k);
  }
}
